tests/Unit/DataStructures/VectorImplTestHelper.hpp

```diff
@@ -366,5 +376,338 @@ void vector_test_math_after_move(
+///
+/// \details Three types of test are provided:
+/// - `Normal` is used to indicate those tests which should be performed over
+///   all combinations of the supplied vector type(s) and their value
+///   types. This is useful for e.g. `+`.
+///
+/// - `Strict` is used to indicate those tests which should be performed over
+///   only sets of the vector type and compared to the same operation of the set
+///   of its value type. This is useful for e.g. `atan2`, which cannot take a
+///   `DataVector` and a double as arguments.
+///
+/// - `Inplace` is used to indicate those tests which should be performed
+///   maintaining the type of the left-hand side of the operator and not
+///   including it in the combinations. Inplace operators such as `+=` have a
+///   more restrictive condition on the type of the left hand side than do
+///   simply `+`. (e.g. `double + complex<double>` compiles, but
+///   `double += complex<double>` does not)
+enum TestKind { Normal, Strict, Inplace };
+
+namespace detail {
+
+// to choose between possible reference wrappers for the math tests.
+namespace UseRefWrap {
+struct Cref {};
+struct None {};
+struct Ref {};
+}  // namespace UseRefWrap
+
+// Wrap is used to wrap values in a std::reference_wrapper using std::cref and
+// std::ref, or to not wrap at all. This is done to verify that all math
+// operations work transparently with a `std::reference_wrapper` too.
+template <class T>
+decltype(auto) wrap(UseRefWrap::Cref /*meta*/, T& t) noexcept {
+  return std::cref(t);
+}
+
+template <class T>
+decltype(auto) wrap(UseRefWrap::None /*meta*/, T& t) noexcept {
+  return t;
+}
+
+template <class T>
+decltype(auto) wrap(UseRefWrap::Ref /*meta*/, T& t) noexcept {
+  return std::ref(t);
+}
+
+using WrapperList =
+    tmpl::list<UseRefWrap::Cref, UseRefWrap::None, UseRefWrap::Ref>;
+
+using NonConstWrapperList = tmpl::list<UseRefWrap::None, UseRefWrap::Ref>;
+
+// struct used for determining the full number of elements in a vector, array of
+// vectors. Designed for use with `test_element_wise_function`
+struct VectorOrArraySize {
+  // array of vectors version, assumes all vectors in the array are the same
+  // size, as will be the case for the relevant test helpers in this detail
+  // namespace
+  template <typename T, size_t S>
+  size_t operator()(const std::array<T, S>& container) const noexcept {
+    return S * VectorOrArraySize{}(container[0]);
+  }
+  // vector version
+  template <typename T,
+            Requires<cpp17::is_arithmetic_v<typename T::ElementType>> = nullptr>
+  size_t operator()(const T& container) const noexcept {
+    return container.size();
+  }
+};
+
+// struct used for obtaining an indexed value in a vector, array of vectors.
+// Designed for use with `test_element_wise_function`
+struct VectorOrArrayAt {
+  // array of vectors version
+  template <typename T, size_t S>
+  auto operator()(std::array<T, S>& container, size_t index) noexcept {
```

Comment:
You want these to have `decltype(auto)` or `T&` as the return type, with the appropriate `const` where necessary if you go for the `T&` approach :)